From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: AlphaKR93 <alphakr93@gmail.com>
Date: Thu, 19 Jan 2023 17:14:08 +0900
Subject: [PATCH] Rebrand


diff --git a/build.gradle.kts b/build.gradle.kts
index 066082881a66af7b202e44d09e2d575842cb7a1d..0b644b49adea7758f2e0ea75f9ed682c9a3e53f5 100644
--- a/build.gradle.kts
+++ b/build.gradle.kts
@@ -160,7 +160,7 @@ publishing {
     repositories {
         maven {
             name = "githubPackage"
-            url = uri("https://maven.pkg.github.com/TeamEarendel/Andromeda")
+            url = uri("https://maven.pkg.github.com/TeamEarendel/AndromedaPlus") // AndromedaPlus
 
             credentials.username = System.getenv("GITHUB_USERNAME")
             credentials.password = System.getenv("GITHUB_TOKEN")
